### ISSUE TITLE
EM_JS code fails when using upstream Clang

### ISSUE THREAD

**Issue by cwoffenden** [emscripten-core/emscripten#8879]
Using the latest upstream Clang and compiling code containing `EM_JS` to a `.bc` then linking errors with `undefined symbol`. The minimum reproducible test comprises two files:
```C++
#include <stdio.h>

extern int test();

int main() {
    if (test()) puts("Hello!");
}
```
```C++
#include <emscripten/em_js.h>

EM_JS(int, testjs, (), {
    return 1;
})

int test() {
    return testjs();
}
```
Putting all the code in one file doesn't fail. Compile with: `emcc main.c test.c -o out.bc` followed by `emcc out.bc -o out.html`. The resulting error is:
```
error: undefined symbol: testjs
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
Error: Aborting compilation due to previous errors
shared:ERROR: 'C:/Emscripten/node/8.9.1_64bit/bin/node.exe C:\Emscripten\emscripten\incoming\src\compiler.js C:\Users\carl\AppData\Local\Temp\tmpe2gumy.txt C:\Emscripten\emscripten\incoming\src\library_pthread_stub.js' failed (1)
```
Compiling directly without going to `.bc` works. Compiling as C++ also fails.

The two files are attached: [test.zip](https://github.com/emscripten-core/emscripten/files/3335789/test.zip).

**Comment by cwoffenden** [comment]
Just add, compiling the `EM_JS` code alone to a `.bc` doesn't fail, e.g.: `emcc test.c -o test.bc` followed by `emcc main.c test.bc -o out.html`.

The error is specifically when compiling all to a `.bc` then creating the wasm from that.

**Comment by jgravelle-google** [comment]
Looked at this with help from @sbc100

Of note is the .bc files as compiled with the upstream backend are, in fact, .o object files. To get actual bitcode, you can compile with `emcc main.c test.c -o out.bc -s WASM_OBJECT_FILES=0`, and this works.

This is why `test.bc` works, because it's really an object file. So that's equivalent to `emcc -c test.c -o test.o && emcc main.c test.o -o out.html`

The problem is a bug in the lld linker when merging two object files into another object file, and dropping the object visibility on the generated `__em_js__testjs` function, which keeps it from being DCE'd by the linker so wasm-emscripten-finalize can generate the metadata so emcc can create the `testjs` function in JS (whew).

Linking two object files together is a really unusual case, and you probably don't want to be doing it by accident. For a quick fix I'd say use `-s WASM_OBJECT_FILES=0` to keep generating actual bitcode. Long-term you'll want to build with `emcc -c main.c test.c` which will build main.o and test.o, which can be linked with `emcc main.o test.o -o out.html`. This is preferable because you get much faster incremental linking.

This sort of confusion makes me think we should tweak the "stopping at bitcode" logic (the actual message here is "stopping at object files"). Either make it always actual bitcode, or handle .bc and .o suffixes separately (e.g. .bc implies WASM_OBJECT_FILES=0 if not otherwise set), or at least only handle the object file suffixes we recognize (`emcc test.c -o test.fish` build a test.fish file that is also a wasm object file).

So tl;dr there are several workarounds, but also a bug in the linker.

**Comment by cwoffenden** [comment]
Thanks for the explanation. We were building to `.bc` with the old fastcomp because it let us half our 20 minute build time (compile the main project to `.bc` with some `.bc` libs for code that doesn't change much, then use that to make separate wasm them asm.js builds). Our current workaround is to just build twice.

With the new upstream Clang I'm assuming we'd build wasm then use `wasm2js` to create the asm.js equivalent. We've not tried this approach yet though.

**Comment by kripken** [comment]
@jgravelle-google - in your last comment it sounds like there's a bug in the linker. Do you know if there are any updates there (and is there a better bug filed for it)?

**Comment by jgravelle-google** [comment]
> The problem is a bug in the lld linker when merging two object files into another object file, and dropping the object visibility on the generated __em_js__testjs function, which keeps it from being DCE'd by the linker so wasm-emscripten-finalize can generate the metadata so emcc can create the testjs function in JS (whew).

Is the issue IIRC. I don't think there's any single bug for it; the general problem is that `wasm-ld a.o b.o -o c.o` is a very weird case that I don't think we have much support for. @sbc100 do you know of any tracking bugs for the linker emitting object files from object files?

**Comment by sbc100** [comment]
I assume you mean with `-r/--relocatable`?  It should work but perhaps we have found a bug here that we need to address.

**Comment by sbc100** [comment]
Oh I see now is the problem that visibility is not preserved with `--relocatable`.    No I don't think we have a bug for that yet.  I'll try to make a minimal repro case and open an llvm bug.

**Comment by sbc100** [comment]
Fix is https://reviews.llvm.org/D67729.  Working on test case for it still
